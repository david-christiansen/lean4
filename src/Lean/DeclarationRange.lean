/-
Copyright (c) 2021 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
import Lean.MonadEnv
import Lean.AuxRecursor
import Lean.ToExpr
import Lean.Data.Lsp.PositionEncodingKind

namespace Lean

/-- A precomputed `character` field as in `Lean.Lsp.Position`. We
need to store this because some LSP clients want us to report the range
in terms of UTF-16, but converting a Unicode codepoint stored in
`Lean.Position` to UTF-16 requires loading and mapping the target
source file, which is IO-heavy.

To avoid a tricky invariant that there's a globally consistent
setting, all three possibilities are cached. -/
structure EncodedPosition where
  utf8 : Nat
  utf16 : Nat
  utf32 : Nat
  deriving Inhabited, DecidableEq, Repr

instance : ToExpr EncodedPosition where
  toExpr p := mkAppN (mkConst ``EncodedPosition.mk) #[toExpr p.utf8, toExpr p.utf16, toExpr p.utf32]
  toTypeExpr := mkConst ``EncodedPosition

def EncodedPosition.get (pos : EncodedPosition) : Lsp.PositionEncodingKind → Nat
  | .utf8 => pos.utf8
  | .utf16 => pos.utf16
  | .utf32 => pos.utf32

/-- Store position information for declarations. -/
structure DeclarationRange where
  pos        : Position
  charLsp    : EncodedPosition
  endPos     : Position
  endCharLsp : EncodedPosition
  deriving Inhabited, DecidableEq, Repr

instance : ToExpr DeclarationRange where
  toExpr r   := mkAppN (mkConst ``DeclarationRange.mk) #[toExpr r.pos, toExpr r.charLsp, toExpr r.endPos, toExpr r.endCharLsp]
  toTypeExpr := mkConst ``DeclarationRange

structure DeclarationRanges where
  range          : DeclarationRange
  selectionRange : DeclarationRange
  deriving Inhabited, Repr

instance : ToExpr DeclarationRanges where
  toExpr r   := mkAppN (mkConst ``DeclarationRanges.mk) #[toExpr r.range, toExpr r.selectionRange]
  toTypeExpr := mkConst ``DeclarationRanges

builtin_initialize builtinDeclRanges : IO.Ref (NameMap DeclarationRanges) ← IO.mkRef {}
builtin_initialize declRangeExt : MapDeclarationExtension DeclarationRanges ← mkMapDeclarationExtension

def addBuiltinDeclarationRanges (declName : Name) (declRanges : DeclarationRanges) : IO Unit :=
  builtinDeclRanges.modify (·.insert declName declRanges)

def addDeclarationRanges [MonadEnv m] (declName : Name) (declRanges : DeclarationRanges) : m Unit :=
  modifyEnv fun env => declRangeExt.insert env declName declRanges

def findDeclarationRangesCore? [Monad m] [MonadEnv m] (declName : Name) : m (Option DeclarationRanges) :=
  return declRangeExt.find? (← getEnv) declName

def findDeclarationRanges? [Monad m] [MonadEnv m] [MonadLiftT IO m] (declName : Name) : m (Option DeclarationRanges) := do
  let env ← getEnv
  let ranges ← if isAuxRecursor env declName || isNoConfusion env declName || (← isRec declName)  then
    findDeclarationRangesCore? declName.getPrefix
  else
    findDeclarationRangesCore? declName
  match ranges with
  | none => return (← builtinDeclRanges.get (m := IO)).find? declName
  | some _ => return ranges

end Lean
