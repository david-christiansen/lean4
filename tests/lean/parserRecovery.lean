import Lean

open Lean Parser

def nonws : Parser where
  fn :=
    andthenFn
      (andthenFn (satisfyFn (fun c => !c.isWhitespace && c != '#')) (takeWhileFn (!·.isWhitespace)))
      whitespace

def thing : Parser :=
  recover ident nonws

def "foo" (x : Nat) :  := 5

def 5 : Nat := 3

theorem yep : True := by trivial

theorem () : Nat := 99
theorem 2.4 : Nat := 99
theorem "nat" : Nat := 99
theorem 3 : Nat := 99

opaque ((( : Nat := )

axiom "nope" : 4 = ⟩

class inductive () where
  | "" : )

inductive 44: Type where | "" : 44

inductive 44 : Type where | "" : 44

inductive 44 (n : Nat) : Type where | "" : 44

inductive 44(n : Nat) : Type where | "" : 44

inductive 44| "" : 44

inductive !!! where
  | "" (x : Nat)) : Bogus

#eval Foo.a

inductive Item.{4, 5, 6, 7, a, "foo", b, c} : Nat where
  | foo : )
  | bar : )

example := ""

-- Adding recovery to Lean's term grammar doesn't work well with the primitive tools at hand
-- (extensibility + token tables + backtracking are a real challenge), but for DSLs it can be nice.

def altParen : Parser :=
  "{|" >> termParser (prec := 51) >> recover "|}" skip

macro:50 e:altParen : term => pure ⟨e.raw[1]⟩

-- These terms show multiple errors due to recovery from missing |} tokens
#eval ([ {|2 + 3] * {| 3 + |}

-- Only one error here
#eval ([ (2 + 3] * ( 3 + ) )

example := ""

open Lean Elab Command in
elab "#go " xs:thing* (Lean.Parser.eoi <|> "#stop") : command => do
  let xs := toString xs
  elabCommand <| ← `(#eval s!"hey {$(quote xs)}")

-- Many errors due to recovery!
#go
hey
5
a
g
def
5
hey
(     99
  (
#stop

def x := 5
#eval x
