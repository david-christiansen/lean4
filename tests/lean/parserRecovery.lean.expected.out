parserRecovery.lean:14:3-14:9: error: unexpected token; expected identifier
parserRecovery.lean:14:21-14:25: error: unexpected token ':='; expected term
parserRecovery.lean:16:3-16:5: error: unexpected token; expected identifier
parserRecovery.lean:20:7-20:9: error: unexpected token '('; expected identifier
parserRecovery.lean:20:9-20:10: error: unexpected token ')'; expected '_' or identifier
parserRecovery.lean:21:7-21:11: error: unexpected token; expected identifier
parserRecovery.lean:21:20-21:22: error: unexpected token ')'; expected term
parserRecovery.lean:22:7-22:13: error: unexpected token; expected identifier
parserRecovery.lean:22:22-22:24: error: unexpected token ')'; expected term
parserRecovery.lean:23:7-23:9: error: unexpected token; expected identifier
parserRecovery.lean:23:18-23:20: error: unexpected token ')'; expected term
parserRecovery.lean:25:6-25:8: error: unexpected token '('; expected identifier
parserRecovery.lean:25:8-25:9: error: unexpected token '('; expected '_' or identifier
parserRecovery.lean:27:5-27:12: error: unexpected token; expected identifier
parserRecovery.lean:27:18-27:20: error: unexpected token '‚ü©'; expected term
parserRecovery.lean:29:15-29:17: error: unexpected token '('; expected identifier
parserRecovery.lean:29:17-29:18: error: unexpected token ')'; expected '_' or identifier
parserRecovery.lean:32:9-32:12: error: unexpected token; expected identifier
parserRecovery.lean:32:27: error: expected token
parserRecovery.lean:34:9-34:12: error: unexpected token; expected identifier
parserRecovery.lean:34:28: error: expected token
parserRecovery.lean:36:9-36:12: error: unexpected token; expected identifier
parserRecovery.lean:36:38: error: expected token
parserRecovery.lean:38:9-38:12: error: unexpected token; expected identifier
parserRecovery.lean:38:37: error: expected token
parserRecovery.lean:40:9-40:12: error: unexpected token; expected identifier
parserRecovery.lean:40:14: error: expected token
parserRecovery.lean:42:9-42:11: error: unexpected token '!'; expected identifier
parserRecovery.lean:43:4: error: expected token
parserRecovery.lean:45:6-45:11: error: unknown identifier 'Foo.a'
parserRecovery.lean:47:16-47:17: error: unexpected token; expected identifier
parserRecovery.lean:47:18-47:20: error: unexpected token; expected identifier
parserRecovery.lean:47:21-47:23: error: unexpected token; expected identifier
parserRecovery.lean:47:24-47:26: error: unexpected token; expected identifier
parserRecovery.lean:47:30-47:36: error: unexpected token; expected identifier
parserRecovery.lean:48:9-48:11: error: unexpected token ')'; expected term
parserRecovery.lean:62:16-62:17: error: unexpected token ']'; expected '|}'
parserRecovery.lean:62:20: error: unexpected token at this precedence level; consider parenthesizing the term
parserRecovery.lean:65:15-65:16: error: unexpected token ']'; expected ')', ',' or ':'
parserRecovery.lean:76:3-77:1: error: unexpected token; expected identifier
parserRecovery.lean:79:1-80:3: error: unexpected token 'def'; expected identifier
parserRecovery.lean:81:0-81:1: error: unexpected token; expected identifier
parserRecovery.lean:82:3-83:1: error: unexpected token '('; expected identifier
parserRecovery.lean:83:6-83:8: error: unexpected token; expected identifier
parserRecovery.lean:84:2-84:3: error: unexpected token '('; expected identifier
5
